src/Microsoft.AspNetCore.Server.Kestrel/Http/SocketInput.cs

```diff
@@ -180,25 +180,29 @@ public MemoryPoolIterator ConsumingStart()
 
             lock (_sync)
             {
-                if (!consumed.IsDefault)
+                // object was disposed
+                if (_head != null)
                 {
-                    returnStart = _head;
-                    returnEnd = consumed.Block;
-                    _head = consumed.Block;
-                    _head.Start = consumed.Index;
-                }
+                    if (!consumed.IsDefault)
```

Comment:
You only really care whether `_head` is null or not in this first `if`. I'm not sure, but I think you should leave the subsequent `if` block outside of the null check to avoid potential hangs. I'll wait for @halter73's input on this.